New arg split-by-locale-to added to translate.js

AdamRamberg opened the pull request:
This PR adds the functionality to `translate.js` to split the translations into separate files based on local. The arg to do so is called 'split-by-locale-to', and I'm open for suggestions on the naming of the arg. This functionality will reduce the manual work of code splitting / lazy load translations based on locale. 

The PR was based on the discussions in [this](https://github.com/facebookincubator/fbt/issues/71) issue. However, the implementation differs a bit from the solution presented [here](https://github.com/facebookincubator/fbt/issues/71#issuecomment-525994553). The reason for the difference is that `translate.js` outputs default to stdout, which would be messy to deal with when we want to output to several files. The solution [here](https://github.com/facebookincubator/fbt/issues/71#issuecomment-525994553) could have been used if a new script was created, however in my opinion it seems like a better option to add an arg instead to the existing script. 

NOTE: I'm probably overseeing something, but I was not able to run any linter or any tests for the project before the commit. Seems like there are no eslint config? And there are only test scripts and a jest config added for the demo app?

jrwats commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -208,5 +218,21 @@ if (argv[args.HELP]) {
+if (splitByLocaleTo) {
+  fs.mkdirSync(splitByLocaleTo, {recursive: true});
+
+  Object.keys(output).forEach(function(locale) {
```
I'm also fine with `for (let locale in output) {...}`, but if you want to keep the functional style, use a lambda: `Object.keys(...).forEach(locale => {...})`

AdamRamberg replied:
Fixed in 27cc3d4

kayhadrin commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -208,5 +218,21 @@ if (argv[args.HELP]) {
-process.stdout.write(json);
+
+function createJSON(obj) {
+  return JSON.stringify(obj, ...(argv[args.PRETTY] ? [null, ' '] : []));
```
nit: use 2 spaces for indentation.

```
  return JSON.stringify(obj, ...(argv[args.PRETTY] ? [null, 2] : []));
```

AdamRamberg replied:
Fixed in 27cc3d4

kayhadrin commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -140,7 +142,15 @@ const argv = yargs
-  .alias(args.HELP, 'help').argv;
+  .alias(args.HELP, 'help')
+  .string(args.SPLIT_BY_LOCALE_TO)
+  .default(args.SPLIT_BY_LOCALE_TO, null)
```
Assuming that the new argument name will be `OUTPUT_DIR`, let's add an alias for it named `o`.

E.g.

`.alias(args.OUTPUT_DIR, 'o')`

AdamRamberg replied:
Fixed in 27cc3d4

kayhadrin commented:
I like where this PR is going!

Could you add a test script in this `package.json` file? Something named like `test-translate-fbts-into-output-dir`...

It's similar to this:
https://github.com/facebookincubator/fbt/blob/fd07a6b20d440d49b013472d50ed6f4ac9cf2594/demo-app/package.json#L15

Finally, please add a new entry in the [CHANGELOG](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md) file for posterity! :-)

AdamRamberg commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -77,14 +77,15 @@
  */
 'use strict';
-
+ 
```
Fixed in 27cc3d412d8bb15b04d84110b6d26555708511a8

AdamRamberg commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -94,6 +95,7 @@ const args = {
   SRC: 'source-strings',
   STDIN: 'stdin',
   TRANSLATIONS: 'translations',
+  SPLIT_BY_LOCALE_TO: 'split-by-locale-to',
```
Fixed in 27cc3d4

AdamRamberg commented on transform/babel-plugin-fbt/bin/translate.js:
```diff
@@ -208,5 +218,21 @@ if (argv[args.HELP]) {
+  return JSON.stringify(obj, ...(argv[args.PRETTY] ? [null, ' '] : []));
+}
+
+const splitByLocaleTo = yargs.argv[args.SPLIT_BY_LOCALE_TO];
```
Fixed in 27cc3d4

AdamRamberg commented:
Fixed all the issues and added the stuff requested in commit 27cc3d412d8bb15b04d84110b6d26555708511a8. 

> Finally, please add a new entry in the CHANGELOG file for posterity! :-)

Not sure if I'm suppose to do this since I don't know the commit hash (you might squash and merge). I also don't have access to publish to npm and I think we should add a minior to [babel-plugin-fbt](https://github.com/facebookincubator/fbt/blob/master/transform/babel-plugin-fbt/package.json) (0.9.14 -> 0.10.0), which I think should be reflected in the [changelog](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md). This work should probably be made of the one in charge of doing the releases for fbt right? 

Let me know if there is anything else that needs to be changed / added! :)

AdamRamberg commented:
Got another question. Should I add the `test-translate-fbts-into-output-dir`script to the [live-demo-app](https://github.com/facebookincubator/fbt/tree/master/live-demo-app)? Not really sure what the purpose of that is compared to the regular [demo-app](https://github.com/facebookincubator/fbt/tree/master/demo-app) 🤔

kayhadrin commented:
> Fixed all the issues and added the stuff requested in commit [27cc3d4](https://github.com/facebookincubator/fbt/commit/27cc3d412d8bb15b04d84110b6d26555708511a8).
> 
> > Finally, please add a new entry in the CHANGELOG file for posterity! :-)
> 
> Not sure if I'm suppose to do this since I don't know the commit hash (you might squash and merge). I also don't have access to publish to npm and I think we should add a minior to [babel-plugin-fbt](https://github.com/facebookincubator/fbt/blob/master/transform/babel-plugin-fbt/package.json) (0.9.14 -> 0.10.0), which I think should be reflected in the [changelog](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md). This work should probably be made of the one in charge of doing the releases for fbt right?
> 
> Let me know if there is anything else that needs to be changed / added! :)

Feel free to add the CHANGELOG entry without the proper git commit hash. You could simply use the github PR number link instead.
I can do that too if you prefer, no worries.

----

> Got another question. Should I add the `test-translate-fbts-into-output-dir`script to the [live-demo-app](https://github.com/facebookincubator/fbt/tree/master/live-demo-app)? Not really sure what the purpose of that is compared to the regular [demo-app](https://github.com/facebookincubator/fbt/tree/master/demo-app) 🤔

There's no need to add it to the "live demo app" because I don't expect it to remain in the repo for much longer. I'll probably archive it in a git tag and remove it from the repo since it's kind of a duplicate of the normal demo app.

kayhadrin commented:
> I think we should add a minior to [babel-plugin-fbt](https://github.com/facebookincubator/fbt/blob/master/transform/babel-plugin-fbt/package.json) (0.9.14 -> 0.10.0), which I think should be reflected in the [changelog](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md). 

@AdamRamberg Sorry but I'm not really sure what you meant here? 🤔

AdamRamberg commented:
> I think we should add a minior to babel-plugin-fbt (0.9.14 -> 0.10.0), which I think should be reflected in the changelog.
> 
> @AdamRamberg Sorry but I'm not really sure what you meant here? 🤔

Since we are adding functionality to `babel-plugin-fbt`(output-dir arg to the `translate.js` script) my thought was that we should probably increase the minior version of `babel-plugin-fbt`before it gets published to npm next time. Currently it's at 0.9.14 and wlll be at 0.10.0 after. This should also be reflected in the [changelog](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md), eg: 

```
babel-plugin-fbt versions
0.9.14:
c11e9fd Enable custom Babel plugins for FBT collection

0.10.0:
#73 New arg `output-dir` added to the translate script. The output translation files will be split by locale. 
```

So should I increase the minior version of `babel-plugin-fbt` and add a comment to the [changelog](https://github.com/facebookincubator/fbt/blob/master/CHANGELOG.md)? :)

kayhadrin commented:
Thanks for the clarification! I understand what you meant now. 
For now, let's edit the CHANGELOG like follow:

```
### babel-plugin-fbt versions

- Next MINOR version
  - #73 New arg `output-dir` added to the translate script. The output translation files will be split by locale.
- 0.9.14:
  - [c11e9fd](https://github.com/facebookincubator/fbt/commit/c11e9fd) Enable custom Babel plugins for FBT collection
```

I would prefer using this format to avoid setting *hard* expectations on the *future* npm version numbers.
I'll take care of updating the CHANGELOG and the version numbers before publishing to npm. :-)

AdamRamberg commented:
I just added an entry to the CHANGELOG.md file. Sorry for the world's slowest copy and paste ;) 

Let me know if there are any other changes required before merging.

kayhadrin commented:
👍
Updated code looks good and my translation tests were successful too.

kayhadrin commented:
@jrwats, I think we'll also need your approval before merging this PR.

jrwats commented:
LGTM, but you've already approved it.  We'll need to bump some internal tools checksums for the changes is all.
